internal/app/orchestrator/orchestrator.go

```diff
@@ -90,16 +110,176 @@ func New(ctx context.Context, l log.Logger, mapper mapper.Mapper, upstreamClient
+	}
+
+	// Register the watch for future responses.
+	err = o.cache.AddRequest(aggregatedKey, &req)
+	if err != nil {
+		// If we fail to register the watch, we need to kill this stream by
+		// closing the response channel.
+		o.logger.With("err", err).With("key", aggregatedKey).With(
+			"req node", req.GetNode()).Error(ctx, "failed to add watch")
+		closedChannel := o.downstreamResponseMap.delete(&req)
+		return closedChannel, nil
+	}
+
+	// Check if we have a cached response first.
+	cached, err := o.cache.Fetch(aggregatedKey)
+	if err != nil {
+		// Log, and continue to propagate the response upstream.
+		o.logger.With("err", err).With("key", aggregatedKey).Warn(ctx, "failed to fetch aggregated key")
+	}
+
+	if cached != nil && cached.Resp != nil && cached.Resp.Raw.GetVersionInfo() != req.GetVersionInfo() {
+		// If we have a cached response and the version is different,
+		// immediately push the result to the response channel.
+		go func() { responseChannel <- convertToGcpResponse(cached.Resp, req) }()
+	}
+
+	// Check if we have a upstream stream open for this aggregated key. If not,
+	// open a stream with the representative request.
+	//
+	// Locking is necessary here so that a simultaneous downstream request
+	// that maps to the same aggregated key doesn't result in two upstream
+	// streams.
+	o.upstreamResponseMap.mu.Lock()
+	if _, ok := o.upstreamResponseMap.responseChannel[aggregatedKey]; !ok {
+		upstreamResponseChan := o.upstreamClient.OpenStream(ctx, &req)
+		respChannel := o.upstreamResponseMap.add(aggregatedKey, upstreamResponseChan)
+		// Spin up a go routine to watch for upstream responses.
+		// One routine is opened per aggregate key.
+		go o.watchUpstream(ctx, aggregatedKey, respChannel.response, respChannel.done)
+
+	}
+	o.upstreamResponseMap.mu.Unlock()
+
+	return responseChannel, o.onCancelWatch(&req)
 }
 
 // Fetch implements the polling method of the config cache using a non-empty request.
-func (c *orchestrator) Fetch(context.Context, discovery.DiscoveryRequest) (*gcp.Response, error) {
+func (o *orchestrator) Fetch(context.Context, discovery.DiscoveryRequest) (*gcp.Response, error) {
 	return nil, fmt.Errorf("Not implemented")
 }
 
-func (c *orchestrator) onCacheEvicted(key string, resource cache.Resource) {
-	// TODO implement.
+// watchResponse is intended to be called in a goroutine, to receive incoming
+// responses and fan out to downstream clients.
+func (o *orchestrator) watchUpstream(
+	ctx context.Context,
+	aggregatedKey string,
+	responseChannel <-chan *upstream.Response,
+	done <-chan bool,
+) {
+	for {
+		select {
+		case x := <-responseChannel:
+			if x.Err != nil {
+				// A problem occurred fetching the response upstream, log and
+				// return the most recent cached response, so that the
+				// downstream will reissue the discovery request.
+				o.logger.With("err", x.Err).With("key", aggregatedKey).Error(ctx, "upstream error")
```

Comment:
should we still progress if an `Err` was received?

Reply:
also i think `OpenStream` should be called again to reinitiate the stream.

Reply:
I'm still operating under the model than an error is not a non-recoverable state for the stream. I think if it's non-recoverable, closing the channel makes more sense. Let's discuss more in our meeting tomorrow.

Reply:
A stream could be overloaded or in a bad state, but opening a new stream in that scenario would be ok, no?